HeterogeneousCore/SonicTriton/src/TritonMemResource.cc

```diff
@@ -0,0 +1,169 @@
+  }
+
+  //close descriptor
+  if (::close(shm_fd) == -1) {
+    triton_utils::warnOrThrow("unable to close descriptor for shared memory key: " + this->name_, canThrow);
+    this->status_ &= false;
+  }
+
+  this->status_ &= triton_utils::warnOrThrowIfError(
+      this->data_->client()->RegisterSystemSharedMemory(this->name_, this->name_, this->size_),
+      "unable to register shared memory region: " + this->name_,
+      canThrow);
+}
+
+template <typename IO>
+TritonCpuShmResource<IO>::~TritonCpuShmResource<IO>() {
+  triton_utils::warnIfError(this->data_->client()->UnregisterSystemSharedMemory(this->name_),
+                            "unable to unregister shared memory region: " + this->name_);
+
+  //unmap
+  int tmp_fd = munmap(this->addr_, this->size_);
+  if (tmp_fd == -1)
+    edm::LogWarning("TritonWarning") << "unable to munmap for shared memory key: " << this->name_;
+
+  //unlink
+  int shm_fd = shm_unlink(this->name_.c_str());
+  if (shm_fd == -1)
+    edm::LogWarning("TritonWarning") << "unable to unlink for shared memory key: " << this->name_;
+}
+
+template <>
+void TritonInputCpuShmResource::copy(const void* values, size_t offset) {
+  std::memcpy(addr_ + offset, values, data_->byteSizePerBatch_);
+}
+
+template <>
+void TritonOutputCpuShmResource::copy(const uint8_t** values) {
+  *values = addr_;
+}
+
+template <typename IO>
+TritonGpuShmResource<IO>::TritonGpuShmResource(TritonData<IO>* data, const std::string& name, size_t size, bool canThrow)
+    : TritonMemResource<IO>(data, name, size, canThrow), deviceId_(0), handle_(std::make_shared<cudaIpcMemHandle_t>()) {
+  this->status_ &= triton_utils::cudaCheck(
+      cudaMalloc((void**)&this->addr_, this->size_), "unable to allocate GPU memory for key: " + this->name_, canThrow);
+  //todo: get server device id somehow?
+  this->status_ &= triton_utils::cudaCheck(
+      cudaSetDevice(deviceId_), "unable to set device ID to " + std::to_string(deviceId_), canThrow);
+  this->status_ &= triton_utils::cudaCheck(
+      cudaIpcGetMemHandle(handle_.get(), this->addr_), "unable to get IPC handle for key: " + this->name_, canThrow);
+  this->status_ &= triton_utils::warnOrThrowIfError(
+      this->data_->client()->RegisterCudaSharedMemory(this->name_, *handle_, deviceId_, this->size_),
+      "unable to register CUDA shared memory region: " + this->name_,
+      canThrow);
+}
+
+template <typename IO>
+TritonGpuShmResource<IO>::~TritonGpuShmResource<IO>() {
+  triton_utils::warnIfError(this->data_->client()->UnregisterCudaSharedMemory(this->name_),
+                            "unable to unregister CUDA shared memory region: " + this->name_);
+  triton_utils::cudaCheck(cudaFree(this->addr_), "unable to free GPU memory for key: " + this->name_, false);
+}
+
+template <>
+void TritonInputGpuShmResource::copy(const void* values, size_t offset) {
+  triton_utils::cudaCheck(
+      cudaMemcpy((void*)(addr_ + offset), values, data_->byteSizePerBatch_, cudaMemcpyHostToDevice),
```

Comment:
When I try `cudaMemcpyAsync` on the test examples, I see worse performance. That seems strange, but maybe it's because the examples are just simple models rather than existing workflows. I propose keeping it synchronous for now and coming back to it when realistic workflows are available.